jwst/barshadow/bar_shadow.py

```diff
@@ -0,0 +1,440 @@
+# a list of Slits, each of which has several data arrays and
+# keyword metadata items associated 
+#
+# At this point we're going to have to assume that a Slit is composed
+# of a set of slit[].nshutters in a vertical line
+#
+# Reference file information is a 1x1 ref file and a 1x3 ref file
+# Both the 1x1 and 1x3 ref files are 1001 pixel high and go from
+# -1 to 1 in their Y value WCS.  
+# 
+# 
+    exp_type = input_model.meta.exposure.type
+    log.info(exp_type)
+    #
+    # Create the pieces that are put together to make the barshadow model
+    shutter_elements = create_shutter_elements(barshadow_model)
+    w0 = barshadow_model.crval1
+    wave_increment = barshadow_model.cdelt1
+    if exp_type == 'NRS_MSASPEC':
+        # For each slitlet
+        for slitlet in input_model.slits:
+            slitlet_number = slitlet.slitlet_id
+            log.info('Working on slitlet %d' % slitlet_number)
+            if has_uniform_source(slitlet):
+                #
+                # As Y increases, the pixel row number decreases, so the shutter_state is
+                # 'upside down'
+                shutter_status = slitlet.shutter_state[::-1]
+                if len(shutter_status) > 0:
+                    shadow = create_shadow(shutter_elements, shutter_status)
+                    #
+                    # For each pixel in the slit subarray
+                    #   Make a grid of indices for pixels in the subarray
+                    x, y = wcstools.grid_from_bounding_box(slitlet.meta.wcs.bounding_box, step=(1,1))
+                    #   Create the transformation from slit_frame to detector
+                    det2slit = slitlet.meta.wcs.get_transform('detector', 'slit_frame')
+                    #   Use this transformation to calculate x, y and wavelength
+                    xslit, yslit, wavelength = det2slit(x, y)
+                    #   Scale the yslit values by the ratio of slit spacing to slit height
+                    yslit = yslit * SLITRATIO
+                    #   Convert the Y and wavelength to a pixel location
+                    #   in the  bar shadow array
+                    index_of_fiducial = shutter_status.find('x')
+                    index_of_fiducial_in_array = 501 + index_of_fiducial*500
+                    yrow = index_of_fiducial_in_array - yslit*500.0
+                    wcol = (wavelength - w0)/wave_increment
+                    nonnan = np.where(~np.isnan(yrow))
+                    ymin = yrow[nonnan].min()
+                    ymax = yrow[nonnan].max()
+                    wmin = wcol[nonnan].min()
+                    wmax = wcol[nonnan].max()
+                    print(ymin, ymax, wmin, wmax, shutter_status)
+                    #   Interpolate the bar shadow correction for non-Nan pixels
+                    correction = interpolate(yrow, wcol, shadow)
+                    #   Divide the data by the correction
+                    slitlet.data = slitlet.data / correction
```

Comment:
Are we supposed to actually be applying the correction to the data within the step or just attaching the correction values to the data model (to be applied at a later time)?

Reply:
The NIRSpec Technical Note ESA-JWST-SCI-NRS-TN-2016-016 says:
"The results can be stored as a new extension in the 2D irregular data product"
I'll take out the line that applies the correction and instead attach the correction to the MultiSpec model.

Reply:
Just comment out the line for now that applies the correction, but definitely also add the correction array as a new extension in the data model. The JCCWG is still trying to reach consensus on whether or not to apply all the corrections as they're computed or apply them later. In either case they want the corrections saved.

Reply:
Added the correction array to the multispec data model.  I removed the line that applies the correction, it's pretty straightforward to put it back if/when needed